### PR TITLE
dnsproxy: add three new features

### PR DESCRIPTION
**Pull request by emi-YFcz5XDj**
**Maintainer:** @1715173329 
**Compile tested:** Apr 29 Snapshot, x86_64 on QEMU
**Run tested:** Apr 29 Snapshot, x86_64 on QEMU

**Description:**
This commit adds the following features:
1. UCI support for local DNS over HTTPS/TLS/QUIC server.
2. UCI support for using private reverse DNS.
3. procd jail with CAP_NET_BIND_SERVICE, allowing dnsproxy to serve on standard ports directly.

**Notes:**
- I left the default values of tls_crt and key empty, since dnsproxy can't read uhttpd's cert and key by default, and I think the generated certs aren't in PEM format either.
- quic-go will attempt to enlarge net.core.rmem_max and its write equivalent, then log an error when it fails. This doesn't affect the functionalities of dnsproxy.

### CODE REVIEW

File: net/dnsproxy/Makefile [hunk 1]
```diff
@@ -49,6 +49,8 @@ define Package/dnsproxy/install
 	$(INSTALL_CONF) $(CURDIR)/files/dnsproxy.config $(1)/etc/config/dnsproxy
 	$(INSTALL_DIR) $(1)/etc/init.d/
 	$(INSTALL_BIN) $(CURDIR)/files/dnsproxy.init $(1)/etc/init.d/dnsproxy
+	$(INSTALL_DIR) $(1)/etc/capabilities/
+	$(INSTALL_DATA) $(CURDIR)/files/dnsproxy.json $(1)/etc/capabilities/dnsproxy.json
```
Comment (1715173329) [inline net/dnsproxy/Makefile:53]: sort it alphabetically, i.e. put it above `$(INSTALL_DIR) $(1)/etc/config/`

File: net/dnsproxy/files/dnsproxy.config [hunk 2]
```diff
@@ -19,6 +19,14 @@ config dnsproxy 'global'
 	option udp_buf_size ''
 	option verbose '0'
 
+config dnsproxy 'tls'
```
Comment (1715173329) [inline net/dnsproxy/files/dnsproxy.config:22]: sort it alphabetically

File: net/dnsproxy/files/dnsproxy.config [hunk 3]
```diff
@@ -29,6 +37,10 @@ config dnsproxy 'cache'
 	option min_ttl ''
 	option max_ttl ''
 
+config dnsproxy 'private_rdns'
```
Comment (1715173329) [inline net/dnsproxy/files/dnsproxy.config:40]: ditto

File: net/dnsproxy/files/dnsproxy.init [hunk 4]
```diff
@@ -66,6 +66,11 @@ load_config_list() {
 
 	is_empty "bogus_nxdomain" "ip_addr" || config_list_foreach "bogus_nxdomain" "ip_addr" "append_param '--bogus-nxdomain'"
 
+	is_enabled "private_rdns" "enabled" && {
```
Comment (1715173329) [inline net/dnsproxy/files/dnsproxy.init:69]: ditto

File: net/dnsproxy/files/dnsproxy.init [hunk 5]
```diff
@@ -78,6 +83,14 @@ load_config_param() {
 	append_param_arg "global" "rate_limit" "--ratelimit"
 	append_param_arg "global" "udp_buf_size" "--udp-buf-size"
 
+	is_enabled "tls" "enabled" && {
```
Comment (1715173329) [inline net/dnsproxy/files/dnsproxy.init:86]: ditto

File: net/dnsproxy/files/dnsproxy.init [hunk 6]
```diff
@@ -102,6 +115,10 @@ start_service() {
 
 	is_enabled "global" "enabled" || return 1
 
+	local crt
+	local key
+	local log
+
```
Comment (1715173329) [inline net/dnsproxy/files/dnsproxy.init:121]:
```suggestion
	local tls_crt tls_key log_file
	config_get tls_crt "tls" "tls_crt"
	config_get tls_key "tls" "tls_key"
	config_get log_file "global" "log_file"
```

File: net/dnsproxy/files/dnsproxy.init [hunk 7]
```diff
@@ -114,6 +131,17 @@ start_service() {
...
+	procd_add_jail dnsproxy ronly log
+	procd_add_jail_mount /etc/ssl/certs/ca-certificates.crt
+	procd_set_param capabilities /etc/capabilities/dnsproxy.json
+
+	config_get crt tls tls_crt
+	procd_add_jail_mount "$crt"
+	config_get key tls tls_key
+	procd_add_jail_mount "$key"
+	config_get log global log_file
+	procd_add_jail_mount_rw "$log"
```
Comment (1715173329) [inline net/dnsproxy/files/dnsproxy.init:143]:
```suggestion
	procd_set_param capabilities "/etc/capabilities/dnsproxy.json"
	procd_add_jail_mount "/etc/ssl/certs/ca-certificates.crt"
	[ -z "$tls_crt" ] || procd_add_jail_mount "$tls_crt"
	[ -z "$tls_key" ] || procd_add_jail_mount "$tls_key"
	[ -z "$log_file" ] || procd_add_jail_mount_rw "$log_file"
```

check whether tls_crt/tls_key/log_file exists

File: net/dnsproxy/files/dnsproxy.json [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+	"ambient": [
+		"CAP_NET_BIND_SERVICE"
+	],
+	"permitted": [
+		"CAP_NET_BIND_SERVICE"
+	],
+	"inheritable": [
+		"CAP_NET_BIND_SERVICE"
+	]
+}
```
Comment (1715173329) [inline net/dnsproxy/files/dnsproxy.json:17]: missing new line (LF)

**Review by 1715173329 — CHANGES_REQUESTED** [review]

**Review by 1715173329 — APPROVED** [review]
LGTM

**Comment by emi-YFcz5XDj** [comment]
Wait there's a trailing tab
Don't merge yet

**Comment by 1715173329** [comment]
Merged, thank you!

**Pull request merged**